src/middlewared/middlewared/plugins/virt/volume.py

```diff
@@ -0,0 +1,119 @@
+        if storage_devices.get('status_code') != 200:
+            return []
+
+        entries = []
+        for storage_device in storage_devices['metadata']:
+            entries.append({
+                'id': storage_device['name'],
+                'name': storage_device['name'],
+                'content_type': storage_device['content_type'].upper(),
+                'created_at': storage_device['created_at'],
+                'type': storage_device['type'],
+                'config': storage_device['config'],
+                'used_by': [instance.replace('/1.0/instances/', '') for instance in storage_device['used_by']]
+            })
+            if storage_device['config'].get('size'):
+                entries[-1]['config']['size'] = int(storage_device['config']['size']) // (1024 * 1024)
+        return filter_list(entries, filters, options)
+
+    @api_method(VirtVolumeCreateArgs, VirtVolumeCreateResult)
+    async def create(self, data):
+        await self.middleware.call('virt.global.check_initialized')
+
+        result = await incus_call('1.0/storage-pools/default/volumes/custom', 'post', {
+            'json': {
+                'name': data['name'],
+                'content_type': data['content_type'].lower(),
+                'config': {
+                    'size': str(data['size'] * 1024 * 1024),  # Convert MB to bytes
+                },
+            },
+        })
+        if result.get('error') != '':
+            raise CallError(f'Failed to create volume: {result["error"]}')
+
+        return await self.get_instance(data['name'])
+
+    @api_method(VirtVolumeUpdateArgs, VirtVolumeUpdateResult)
+    async def do_update(self, name, data):
+        volume = await self.get_instance(name)
+        if data.get('size') is None:
+            return volume
+
+        result = await incus_call(f'1.0/storage-pools/default/volumes/custom/{name}', 'patch', {
+            'json': {
+                'config': {
+                    'size': str(data['size'] * 1024 * 1024)
+                },
+            },
+        })
+        if result.get('error') != '':
+            raise CallError(f'Failed to update volume: {result["error"]}')
+
+        return await self.get_instance(name)
+
+    @api_method(VirtVolumeDeleteArgs, VirtVolumeDeleteResult)
+    async def do_delete(self, name):
+        volume = await self.get_instance(name)
+        if volume['used_by']:
+            raise CallError(f'Volume {name!r} is in use by instances: {", ".join(volume["used_by"])}')
+
+        result = await incus_call(f'1.0/storage-pools/default/volumes/custom/{name}', 'delete')
+        if result.get('status_code') != 200:
+            raise CallError(f'Failed to delete volume: {result["error"]}')
+
+        return True
+
+    @api_method(VirtVolumeImportISOArgs, VirtVolumeImportISOResult)
+    @job(lock=lambda args: f'virt_volume_import_iso_{args[0]}')
+    async def import_iso(self, job, data):
```

Comment:
That means we can only import an ISO that exists in the pool? How about we allow it to be uploaded through the job?